python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/__init__.py

```diff
@@ -0,0 +1,50 @@
+import logging
+from importlib import import_module
+from typing import Collection
+from wrapt import wrap_function_wrapper
+
+from openinference.instrumentation.haystack.version import __version__
+from openinference.instrumentation.haystack._wrappers import _ComponentWrapper, _PipelineWrapper
+from opentelemetry.trace import get_tracer
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry import trace as trace_api
+
+logger = logging.getLogger(__name__)
+
+_instruments = ("haystack-ai >= 2.0.0",)
+
+class HaystackInstrumentor(BaseInstrumentor):
+    """An instrumentor for the Haystack framework."""
+
+    __slots__ = (
+        "_original_pipeline_run",
+    )
+
+    def instrumentation_dependencies(self) -> Collection[str]:
+        return _instruments
+
+    def _instrument(self, **kwargs):
+        if not (tracer_provider := kwargs.get("tracer_provider")):
+            tracer_provider = trace_api.get_tracer_provider()
+        tracer = get_tracer(__name__, __version__, tracer_provider)
+        haystack = import_module("haystack.core.pipeline.pipeline")
+        self._original_pipeline_run = haystack.Pipeline.run
+        self._original_pipeline_run_component = haystack.Pipeline._run_component
+
+        # Creating a parent span for the Pipeline
+        wrap_function_wrapper(
+            module = "haystack.core.pipeline.pipeline",
+            name = "Pipeline.run",
+            wrapper = _PipelineWrapper(tracer=tracer),
+        )
+        # Creating child spans for every Component in the Pipeline
+        wrap_function_wrapper(
+            module = "haystack.core.pipeline.pipeline",
+            name = "Pipeline._run_component",
+            wrapper = _ComponentWrapper(tracer=tracer),
+        )
+
+    def _uninstrument(self, **kwargs):
```

Comment:
missing uninstruments here

Reply:
you also want to add logic to check that these functions were indeed wrapped by us before uninstrumenting

check out this example
https://github.com/Arize-ai/openinference/pull/622/files#diff-743fb0121267b7d5919416f1372ebc7d063a9c89cc2b7ae64ccceb3df519c9a6R61-R75